feat: Automatic node sharing for ConstraintProvider (Enterprise) (#685)
TimefoldAI/timefold-solver: docs/src/modules/ROOT/pages/enterprise-edition/enterprise-edition.adoc

unless there is an OS specific policy in place to avoid Timefold Solver from hogging all the CPU processors.
====

[#automaticNodeSharing]
=== Automatic node sharing

[NOTE]
====
This feature is a commercial feature of Timefold Solver Enterprise Edition.
It is not available in the Community Edition.
====

When a `ConstraintProvider` does an operation for multiple constraints (such as finding all shifts corresponding to an employee), that work can be shared.

> This can massively improve score calculation speed if the repeated operation is computationally expensive.

Reviewer: ```suggestion
This can significantly improve score calculation speed if the repeated operation is computationally expensive.
```

==== Configuration

[tabs]
======
Plain Java::

* Add `<constraintStreamAutomaticNodeSharing>true</constraintStreamAutomaticNodeSharing>` in your `solverConfig.xml`:
+
[source,xml,options="nowrap"]
----
<!-- ... -->
<scoreDirectorFactory>
  <constraintProviderClass>org.acme.MyConstraintProvider</constraintProviderClass>
  <constraintStreamAutomaticNodeSharing>true</constraintStreamAutomaticNodeSharing>
</scoreDirectorFactory>
<!-- ... -->
----

Spring Boot::

Set the property `timefold.solver.constraint-stream-automatic-node-sharing` to `true` in `application.properties`:
+
[source,properties,options="nowrap"]
----
timefold.solver.constraint-stream-automatic-node=true
----

Quarkus::

Set the property `quarkus.timefold.solver.constraint-stream-automatic-node-sharing` to `true` in `application.properties`:
+
[source,properties,options="nowrap"]
----
quarkus.timefold.solver.constraint-stream-automatic-node-sharing=true
----
======

> IMPORTANT: Debugging breakpoints put inside your constraints will not be respected, because the ConstraintProvider class will be transformed when this feature is enabled.

Reviewer: ```suggestion
[IMPORTANT]
====
Debugging breakpoints put inside your constraints will not be respected, because the `ConstraintProvider` class will be transformed when this feature is enabled.
====
```

==== What is node sharing?

When using xref:constraints-and-score/score-calculation.adoc#constraintStreams[constraint streams], each xref:constraints-and-score/score-calculation.adoc#constraintStreamsBuildingBlocks[building block] forms a node in the score calculation network.

> When two building blocks are functionality equivalent, they can share the same node in the network.

Reviewer: ```suggestion
When two building blocks are functionally equivalent, they can share the same node in the network.
```

Sharing nodes allows the operation to be performed only once instead of multiple times, improving the performance of the solver.

> To be functionality equivalent, the following must be true:

Reviewer: ```suggestion
To be functionally equivalent, the following must be true:
```

* The building blocks must represent the same operation.

> * The building blocks must have functionality equivalent parent building blocks.

Reviewer: ```suggestion
* The building blocks must have functionally equivalent parent building blocks.
```

> * The building blocks must have functionality equivalent inputs.

Reviewer: ```suggestion
* The building blocks must have functionally equivalent inputs.
```

> For example, the building blocks below are functionality equivalent:

Reviewer: ```suggestion
For example, the building blocks below are functionally equivalent:
```

[source,java,options="nowrap"]
----
Predicate<Shift> predicate = shift -> shift.getEmployee().getName().equals("Ann");

var a = factory.forEach(Shift.class)
               .filter(predicate);

var b = factory.forEach(Shift.class)
               .filter(predicate);
----

> Whereas these building blocks are not functionality equivalent:

Reviewer: ```suggestion
Whereas these building blocks are not functionally equivalent:
```

[source,java,options="nowrap"]
----
Predicate<Shift> predicate1 = shift -> shift.getEmployee().getName().equals("Ann");
Predicate<Shift> predicate2 = shift -> shift.getEmployee().getName().equals("Bob");

// Different parents
var a = factory.forEach(Shift.class)
               .filter(predicate2);

var b = factory.forEach(Shift.class)
               .filter(predicate1)
               .filter(predicate2);

// Different operations
var a = factory.forEach(Shift.class)
               .ifExists(Employee.class);

var b = factory.forEach(Shift.class)
               .ifNotExists(Employee.class);

// Different inputs
var a = factory.forEach(Shift.class)
               .filter(predicate1);

var b = factory.forEach(Shift.class)
               .filter(predicate2);
----

> Counterintuitively, the building blocks produced by these (seemly) identical methods are not necessarily functionality equivalent:

Reviewer: ```suggestion
Counterintuitively, the building blocks produced by these (seemly) identical methods are not necessarily functionally equivalent:
```

[source,java,options="nowrap"]
----
UniConstraintStream<Shift> a(ConstraintFactory constraintFactory) {
    return factory.forEach(Shift.class)
                  .filter(shift -> shift.getEmployee().getName().equals("Ann"));
}

UniConstraintStream<Shift> b(ConstraintFactory constraintFactory) {
    return factory.forEach(Shift.class)
                  .filter(shift -> shift.getEmployee().getName().equals("Ann"));
}
----

> The Java Virtual Machine is free to (and often does) create different instances of functionality equivalent lambdas.

Reviewer: ```suggestion
The Java Virtual Machine is free to (and often does) create different instances of functionally equivalent lambdas.
```

This severely limits the effectiveness of node sharing, since the only way to know two lambdas are equal is to compare their references.

When automatic node sharing is used, the `ConstraintProvider` class is transformed so all lambdas are accessed via a static final field.

> So a class that looks like this:

Reviewer: ```suggestion
Consider the following input class:
```

[source,java,options="nowrap"]
----
public class MyConstraintProvider implements ConstraintProvider {

    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
        return new Constraint[] {
            a(constraintFactory),
            b(constraintFactory)
        };
    }

    Constraint a(ConstraintFactory constraintFactory) {
        return factory.forEach(Shift.class)
                      .filter(shift -> shift.getEmployee().getName().equals("Ann"))
                      .penalize(SimpleScore.ONE)
                      .asConstraint("a");
    }

    Constraint b(ConstraintFactory constraintFactory) {
        return factory.forEach(Shift.class)
                      .filter(shift -> shift.getEmployee().getName().equals("Ann"))
                      .penalize(SimpleScore.ONE)
                      .asConstraint("b");
    }
}
----

> is transformed to a class that looks like this:

Reviewer: When automatic node sharing is enabled, the class will be transformed to look like this:

> is transformed to a class that looks like this:

Reviewer: ```suggestion
When automatic node sharing is enabled, the class will be transformed to look like this:
```
